### PR TITLE
ImmutableUtil Store's setState should merge state

### PR DESCRIPTION
**Pull request by droberts84**
Currently when using `ImmutableUtil` to use Immutable data in a Store, calling `setState` replaces the current state with the argument passed, which is unexpected as it does not behave like `setState` in React components.

This PR changes `setState` to merge the new provided state with the current state.

There are a few edge cases that become odder with this change, such as removing a key from an `Immutable.Map` (see tests), but overall I think it is a win and much less confusing to users.

Note that this could potentially be a breaking change to current users of `ImmutableUtil`.

### CODE REVIEW

**Comment by mull** [comment]
I wrote my own decorator to do exactly this. I agree with the behaviour. It's a breaking change though of course since it forces you to use Immutable.Map for your state instead of a normal JS object.

**Comment by droberts84** [comment]
I considered doing a check to see if the `merge` function exists before calling it, which would allow you to use a normal JS object in your state.. but I wasn't sure if it was worth the potential performance hit, plus it could still be a breaking change (except without an error) if they were expecting the state to be fully replaced, as it did in previous versions.

**Comment by cayuu** [comment]
I too currently run a monkey patched decorator to get this behaviour. However this current PR will pass through .merge mashing of lists: https://github.com/facebook/immutable-js/issues/406 - the decorator I use applies the `.mergeWith` "fix" in the first comment on that issue. Given that Immutable looks like they're adopting the fix in the next major version bump, I'd suggest to maybe ponyfill the fix or document the hell out of the current wonky List overwriting.

**Comment by droberts84** [comment]
Interesting.  I don't typically use Lists so I wasn't aware of that
behavior, but that definitely is not desired.  That said, I'm hesitant to
add code in Alt to work around an issue with a specific Immutable structure
(especially since they plan to fix it in the next major release).

What do others think?

File: docs/utils/immutable.md [hunk 1]
```diff
@@ -30,9 +30,10 @@ class TodoStore {
   static displayName = 'TodoStore'
 
   constructor() {
-    this.state = {
-      todos: Immutable.Map({})
-    };
+    this.state = Immutable.Map({
+      todos: Immutable.List([]),
+      loading: false
```
Comment (goatslacker) [inline docs/utils/immutable.md]: whats this loading all about?

Comment (goatslacker) [inline docs/utils/immutable.md]: just as an example?

File: src/utils/ImmutableUtil.js [hunk 2]
```diff
@@ -3,7 +3,7 @@ import Immutable from 'immutable'
 function immutable(StoreModel) {
   StoreModel.config = {
     setState(currentState, nextState) {
-      this.state = nextState
+      this.state = currentState.merge(nextState)
```
Comment (goatslacker) [inline src/utils/ImmutableUtil.js]: this works well but then limits us to only being able to use Map

cc @jareware

**Comment by jareware** [comment]
This would work, and it has the nice property of maintaining whatever type state already was.

Still, e.g. #411 actually prevents you from using anything other than a Map, so at least until that's fixed it's a bit of a moot point. :)

File: docs/utils/immutable.md (hunk 1 above)
Comment (droberts84) [inline docs/utils/immutable.md]: Yes `loading` was just meant to be an example of a random variable you might include in state.  Any suggestions to clarify this?

**Comment by goatslacker** [comment]
I'd like to be able to use something other than Map. #485 handles all cases
